lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -0,0 +1,396 @@
+        let dirty = false;
+        let context = context.clone();
+        let jobs = default();
+        let performance = web::window.performance_or_panic();
+        let logger = Logger::new("Shader Compiler");
+        Self { dirty, context, jobs, performance, logger }
+    }
+
+    fn submit<F: 'static + FnOnce(shader::Program)>(
+        &mut self,
+        input: shader::Code,
+        on_ready: F,
+    ) -> JobHandler {
+        self.dirty = true;
+        let strong_handler = JobHandler::new();
+        let handler = strong_handler.downgrade();
+        let on_ready = Box::new(on_ready);
+        let job = Job { input, handler, on_ready };
+        self.jobs.compile.push(job);
+        strong_handler
+    }
+
+    #[profile(Debug)]
+    fn run(&mut self, time: animation::TimeInfo) {
+        if self.dirty {
+            self.run_khr_completion_check_jobs();
+            while self.dirty {
+                match self.run_step() {
+                    Ok(_) => {
+                        if !self.dirty {
+                            break;
+                        }
+                        let now = self.performance.now() as f32;
+                        let current_frame_time =
+                            now - time.animation_loop_start - time.since_animation_loop_started;
+                        if current_frame_time > MAX_FRAME_TIME_MS {
+                            let msg1 =
+                                "Shaders compilation takes more than the available frame time.";
+                            let msg2 = "To be continued in the next frame.";
+                            debug!(self.logger, "{msg1} {msg2}");
+                            break;
+                        }
+                    }
+                    Err(err) => {
+                        if self.context.is_context_lost() {
+                            break;
+                        }
+                        let err_msg = err.blocking_report(&self.context);
+                        error!(self.logger, "{err_msg}");
+                    }
+                }
+            }
+        }
+    }
+
+    #[profile(Detail)]
+    fn run_step(&mut self) -> Result<(), Error> {
+        match () {
+            _ if !self.jobs.compile.is_empty() => self.run_next_compile_job()?,
+            _ if !self.jobs.link.is_empty() => self.run_next_link_job()?,
+            _ if !self.jobs.link_check.is_empty() => self.run_next_link_check_job()?,
+            _ =>
+                if self.jobs.khr_completion_check.is_empty() {
+                    debug!(self.logger, "All shaders compiled.");
+                    self.dirty = false;
+                },
+        }
+        Ok(())
+    }
+
+    #[profile(Detail)]
+    fn run_khr_completion_check_jobs(&mut self) {
+        debug!(self.logger, "Running KHR parallel shader compilation check job.");
+        let khr = self.context.extensions.khr_parallel_shader_compile.as_ref().unwrap();
```

Comment:
The `unwrap` should be an `except` with explanation.

Reply:
You are right, this is implemented in a bad way. I implemented it without this line at all now.